gsoft-inc/wl-authentication-clientcredentialsgrant
# [IDP-1955] Adding logging and tracing for better troubleshooting

## Description of changes

The objectif of this PR is to improve troubleshooting related to token retrieval during client credentials authentication.

- Added logs at different places to understand if we use the cache properly, how long we cache the token, etc.
- Added activities to help searching M2M-related spans in distributed traces.
- Revamp of the integration test to simulate a more realistic distributed environment to help understand how we interact with L1/L2 cache.

The integration test revamp already revealed that the token background acquisition/refresh is a bit too agressive, it doesn't leverage the L2 cache when there's a token that still has a long lifetime.

It is best to review the code of the integration test in the IDE.

Logs can be enabled by setting the default level for the category Workleap.Extensions.Http.Authentication.ClientCredentialsGrant to Debug. For instance they can be published using the OpenTelemetry exporter:

```json
    "OpenTelemetry": {
      "LogLevel": {
        "Default": "Warning",
        "Workleap.Extensions.Http.Authentication.ClientCredentialsGrant": "Debug"
      }
    }
```

## Breaking changes

None

## QA screenshots

Authenticating an HTTP request with a token already in cache (no call to the identity provider is made)
![image](https://github.com/user-attachments/assets/92ad7c64-a2ae-42db-87f1-b986b643f563)

Authenticating an HTTP request - acquire a new token (a POST call is made - I had to disable the background token refresh to see it):
![image](https://github.com/user-attachments/assets/1e7ef7a2-ac4d-45e8-ae47-dde7388bafb3)

Background token refresh - detached span:
![image](https://github.com/user-attachments/assets/d7876dc3-87ed-4629-a7ca-b7cb2ce1f459)

New logs (simulated with a token that has a 30 seconds TTL and cache eviction 10 seconds prior to its expiration)
![image](https://github.com/user-attachments/assets/b65aeb51-6bd1-43c2-849f-983df341a195)
